usr/src/uts/common/fs/zfs/dsl_dir.c

```diff
@@ -1061,9 +1165,31 @@ dsl_dir_space_available(dsl_dir_t *dd,
 	mutex_enter(&dd->dd_lock);
 	if (dsl_dir_phys(dd)->dd_quota != 0)
 		quota = dsl_dir_phys(dd)->dd_quota;
+
+	/*
+	 * Prototype to return the min space avail if lquota is set
+	 */
+	objset_t *os = dd->dd_pool->dp_meta_objset;
+	if (dsl_dir_is_zapified(dd) &&
+	    zap_contains(os, dd->dd_object, DD_FIELD_LOGICALQUOTA) == 0) {
+		VERIFY0(zap_lookup(os, dd->dd_object,
+		    DD_FIELD_LOGICALQUOTA, 8, 1, &logicalquota));
+	}
+
+	if (logicalquota == 0)
+		logicalquota = UINT64_MAX;
+
 	used = dsl_dir_phys(dd)->dd_used_bytes;
-	if (!ondiskonly)
+	logicalused = dsl_dir_phys(dd)->dd_uncompressed_bytes;
+	if (!ondiskonly) {
 		used += dsl_dir_space_towrite(dd);
+		/*
+		 * To check :
+		 * Can't predict how many uncompressed bytes to write
+		 * if compression is enable so arbitrary add bytes to write
```

Comment:
I don't understand the comment.  It's true that dsl_dir_space_towrite() doesn't take into account compression, but that would seem to make it *more* accurate to apply to `logicalused` than to `used`, which we're already doing.  This is a huge overestimate anyway.